lib/endpoints/servers.js

```diff
@@ -1595,6 +1600,196 @@ Server.nop = function handlerServerNop(req, res, next) {
+                    'params must be provided'));
+                return;
+            }
+
+            if (['stage', 'activate'].indexOf(action) === -1) {
+                cb(new errors.PreconditionFailedError(
+                    'Invalid "action" param. It must be one of "stage" or ' +
+                    '"activate"'));
+                return;
+            }
+
+            if (action === 'stage' &&
+                (!req.params.template || !req.params.token)) {
+                cb(new errors.PreconditionFailedError(
+                    '"template" and "token" params must be provided for ' +
+                    'recovery configuration "stage"'));
+                return;
+            }
+
+            ctx.params = {
+                action: action,
+                recovery_uuid: req.params.recovery_uuid,
+                pivtoken: req.params.pivtoken
+            };
+
+            if (req.params.template) {
+                ctx.params.template = req.params.template;
+            }
+
+            if (req.params.token) {
+                ctx.params.token = req.params.token;
+            }
+
+            cb();
+        },
+        function getCnAgentVersion(ctx, cb) {
+            req.stash.server.getFinal({extras: {
+                sysinfo: true,
+                vms: false,
+                disk: false,
+                agents: true,
+                last_heartbeat: false,
+                memory: false
+            }}, function (err, server) {
+                var agent;
+                var idx;
+
+                if (err) {
+                    cb(new VError(err, 'could not determine cn-agent version'));
+                    return;
+                }
+                ctx.zpoolEncrypted = server.sysinfo['Zpool Encrypted'] || false;
+                ctx.zpoolRecovery = server.sysinfo['Zpool Recovery'] || {};
+                ctx.cnAgent = {}; // To ensure we always have an object.
+                for (idx = 0; idx < server.agents.length; idx++) {
+                    agent = server.agents[idx];
+                    if (agent.name === 'cn-agent') {
+                        ctx.cnAgent = agent;
+                        break;
+                    }
+                }
+
+                cb();
+            });
+        },
+        function bailIfCnAgentTooOld(ctx, cb) {
+            if (!ctx.cnAgent.version ||
+                semver.lt(ctx.cnAgent.version,
+                    CN_AGENT_VERSION_WITH_RECOVERY_CONFIG)) {
+                cb(new errors.PreconditionFailedError(
+                    'cn-agent (v%s) on server %s does not support '
+                    + '"recovery_config": require at least cn-agent v%s',
+                    ctx.cnAgent.version, req.stash.server.uuid,
+                    CN_AGENT_VERSION_WITH_RECOVERY_CONFIG));
+                return;
+            }
+            self.log.debug(
+                'server %s cn-agent supports "agents_uninstall": v%s',
```

Comment:
Heh - don't think this is "agents_uninstall".

Reply:
/me laughs